JSON extension

lnkuiper opened the pull request:
This PR contains the initial implementation of the JSON extension.

We implement two functions that operate on strings, which are `json_extract_<type>` and `json_type`, which are similar to SQLite's JSON1 extension. We intend to improve support further down the road.

Some examples of the extract function:
```sql
select json_extract_string('{"my_field": "quack"}', 'my_field');
-- quack
select json_extract_string('{"my_field": "quack"}', 'my_field2');
-- NULL
select json_extract_int('{"my_field": "quack"}', 'my_field');
-- NULL
select json_extract_string('{"my_field": {"my_nested_field": ["goose", "duck"]}}', '/my_field/my_nested_field/1');
-- duck
select json_extract_string('{"my_field": {"my_nested_field": ["goose", "duck"]}}', '$.my_field.my_nested_field[0]');
-- goose
select json_extract_int('[1, 2, 42]', 2);
-- 42
```

Users need to know their JSON schema (i.e. know the types) in order to extract them, as we do not support JSON to JSON extraction (yet). The types can be checked with `json_type`:
```sql
select json_type('{"str": 42}');
-- object
select json_type('[1, 2, 3]');
-- array
select json_type('"other"');
-- NULL

select json_type('{"my_field": "quack"}', 'my_field');
-- string
select json_type('{"my_field": 42}', 'my_field');
-- integer
select json_type('{"my_field": 0.42}', 'my_field');
-- real
select json_type('{"my_field": null}', 'my_field');
-- null
select json_type('{"my_field": true}', 'my_field');
-- boolean
select json_type('{"my_field": {"my_nested_field": ["goose", "duck"]}}', 'my_field');
-- object
select json_type('{"my_field": {"my_nested_field": ["goose", "duck"]}}', '/my_field/my_nested_field');
-- array
```

The string `'"other"' is not valid JSON, therefore we cannot parse it to get the type.

The JSON extension should be installable in the same way other extension are installed soon(tm).

Happy to receive any feedback!

Commit:
humble beginnings json extension

Commit:
proper null handling for JSON extract

Commit:
merge master

Commit:
JSON extension: more tests and add to github actions

Commit:
Merge branch 'master' into yyjson

Commit:
minor nitpicks

Commit:
add json_type function and fix extension installation

Commit:
Merge branch 'master' into yyjson

Commit:
add unary json_type function

Commit:
change how json functions are registered

Commit:
Merge branch 'master' into yyjson

Commit:
fix json_extract_bigint

Commit:
automatically cast path to string

Mytherin commented on extension/json/json_extract.cpp:
```diff
@@ -0,0 +1,142 @@
+	AddFunctionAliases(functions, {"json_extract_bigint"}, bigint_fun);
+	AddFunctionAliases(functions, {"json_extract_ubigint"}, ubigint_fun);
+	AddFunctionAliases(functions, {"json_extract_double"}, double_fun);
+	AddFunctionAliases(functions, {"json_extract_string", "json_extract_varchar"}, string_fun);
```
Perhaps for a future PR: Most systems I have looked at also provide a generic `json_extract` function that itself returns a `json` object again, so it can be used to extract any type.

For example in SQLite:

```sql
create table json(j);
insert into json values ('{"val": "quack"}'), ('{"val": []}'), ('{"val": {"value": 1}}');
select json_extract(j, '$.val') from json;
-- quack                   
-- []                      
-- {"value":1}   
```

Or in MySQL:

```sql
create table json(j json);
insert into json values ('{"val": "quack"}'), ('{"val": []}'), ('{"val": {"value": 1}}');
select json_extract(j, '$.val') from json;
-- quack                   
-- []                      
-- {"value":1}   
```

In Postgres:

```sql
create table json(j json);
insert into json values ('{"val": "quack"}'), ('{"val": []}'), ('{"val": {"value": 1}}');
select j->'val' from json;
-- quack                   
-- []                      
-- {"value":1}   
```

In Postgres, the differentiation is made where `->` returns a JSON object, and `->>` returns a VARCHAR object. Both can be used to extract any type, but `->>` returns a VARCHAR object which can then be cast to any type that is desirable (e.g. we can cast to integer).

lnkuiper replied:
I've implemented a generic `json_extract` :) Going to do the `->` syntax later and add a `JSON` LogicalType

Mytherin commented on test/sql/json/test_json_type.test:
```diff
@@ -0,0 +1,155 @@
+
+# this does not work as it is not valid JSON
+query T
+select json_type('"other"')
```
Perhaps throwing an error is better in this case, e.g. in SQLite:

```sql
sqlite> select json_type('XXX');
Error: malformed JSON
```

Mytherin replied:
I think the end goal should be to implement all of [SQLite's json functions](https://www.sqlite.org/json1.html).

lnkuiper replied:
I've implemented more of SQLite's json functions, still more to go

Mytherin commented on test/sql/json/test_json_extract.test:
```diff
@@ -0,0 +1,193 @@
+
+require json
+
+statement ok
```
Perhaps we could have a look at converting the [SQLite JSON tests](https://github.com/sqlite/sqlite/blob/master/test/json101.test) (sadly not in SQLLogicTest format), especially as we add more of the SQLite JSON functionality.

lnkuiper replied:
I've added some of the tests, will add more as I add more functionality

Mytherin commented on extension/json/include/json_functions.hpp:
```diff
@@ -0,0 +1,26 @@
+class JSONFunctions {
+public:
+	static vector<CreateScalarFunctionInfo> GetFunctions() {
+		auto functions = GetExtractFunctions();
```
This should probably be moved into a `.cpp` file, I suspect this code being in a header is related to the linking failures.

Mytherin commented on extension/json/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+add_library(json_extension STATIC json-extension.cpp json_extract.cpp
+                                  json_type.cpp ${YYJSON_OBJECT_FILES})
+
+build_loadable_extension(json json-extension.cpp ${YYJSON_OBJECT_FILES})
```
This should include all files related to the extension (i.e. including `json_extract.cpp` and `json_type.cpp`). I suspect this is causing the CI issues. I recommend moving the files into a variable similar to how this is done for the other extensions.

lnkuiper replied:
Thanks! I've changed, this, still have to fix more CI stuff though before everything's green

Mytherin reviewed (commented):
Thanks for the PR! Looks great. There seems to be some problems remaining in the extension installation CI runs. Otherwise some feedback:

Commit:
add json_valid and sqlite tests

Commit:
merge master

hannesmuehleisen commented:
CC @zenazn

Commit:
JSON $ syntax working

Commit:
merge master

Commit:
bugfix dollar notation and fix python config (yyjson)

Commit:
throw appropriate JSON errors

Commit:
move json functions to separate folder

Commit:
Merge branch 'master' into yyjson

Commit:
formatting

Commit:
proper exceptions for json dollar path, and start on json_array_length

Commit:
Merge branch 'master' into yyjson

Commit:
add a ton of SQLite tests

lnkuiper commented:
Ah, I finally have more time to do programming. Thanks for the feedback Alex and Mark! I am now following SQLite's extension, aiming to add all functionality that they have. I'm also interested in Postgres's operators, will probably add them too. This won't make the next release, but the one after that should be packed with JSON features!

lnkuiper commented:
Only the Windows extension tests are failing now, and I'm not quite sure how to decipher the error output there. Any guesses?

Commit:
add JSON LogicalType

Commit:
Merge branch 'master' into yyjson

Commit:
add -> json extraction to parser (conflicted with lambda)

Commit:
add list version of JSON functions (e.g. extract multiple)

Commit:
Merge branch 'master' into yyjson

Commit:
add list arg versions of the typed json extract functions

Mytherin commented:
> Only the Windows extension tests are failing now, and I'm not quite sure how to decipher the error output there. Any guesses?

Linking with Windows extensions is a bit problematic, see [here](https://github.com/duckdb/duckdb/pull/2569#issuecomment-981148964) for some text on what I had to do before to solve this.

Generally the issue is either missing `DUCKDB_API` flags in functions, functions defined in headers that should be defined in source files, or using static variables/data instead of functions from within the extension.

Mytherin commented on src/include/duckdb/common/types.hpp:
```diff
@@ -365,7 +365,8 @@ enum class LogicalTypeId : uint8_t {
 	TABLE = 103,
-	ENUM = 104
+	ENUM = 104,
+	JSON = 105
```
Since this is adding a new logical type, could we add it to the `test_all_types` function (`src/function/table/system/test_all_types.cpp`)?

Mytherin commented:
@lnkuiper could you also have a look at resolving the merge conflicts?

lnkuiper commented:
@Mytherin I am going a bit crazy on this extension, probably not ready for merging yet soon because I am adding more functionality. Would you prefer if I withdraw the PR?

Commit:
add JSON logicaltype to test_all_types.cpp, and some initial work on json_object()

Commit:
Merge branch 'master' into yyjson

Commit:
fix some templating issues by moving stuff to the header

Mytherin commented:
> @Mytherin I am going a bit crazy on this extension, probably not ready for merging yet soon because I am adding more functionality. Would you prefer if I withdraw the PR?

If you are still actively working on it and don't want me to review the code, then feel free to withdraw it for now and re-open when you want me to have a look at it again.

Commit:
add cost of 1 for casting varchar/json to prevent conflicts

Commit:
json_object(), json_array(), and json() refactored and working

Commit:
NULLS are now properly set in our struct/map to json conversion

Commit:
properly add JSON to test_all_types

Commit:
add JSON to parquet writer

Commit:
add JSON logical type to some more places

Commit:
add JSON logical type to even more places

Commit:
fix yyjson bug, NULL cast duckdb bug, and add more json test cases

Commit:
add a bunch of sqlite tests

Commit:
merge master

Commit:
add clear error when DUCKDB_EXTENSION_MAIN is not defined because the error is very unclear otherwise

Commit:
pull in new version of yyjson that bugfixed some stuff we need

Commit:
fix workflow typo and fix null cast issue for constant vectors as well

Commit:
properly cast NULL list to varchar

Commit:
fix wrongly merged test

lnkuiper commented:
This PR is now ready for review!

Commit:
implement json_structure function

Commit:
add another json_array test, make json_structure more flexible, start on json_transform

Commit:
fix decimal cast issue and use our LogicalType instead of json types (eventually for implicit casting)

Commit:
add type casting to json_transform

Commit:
merge master

Commit:
merge master

Commit:
Merge branch 'yyjson2' into yyjson

Commit:
add/subtract tiny amount depending on sign

Commit:
some renames for clarity

Commit:
some performance improvements

Commit:
Merge branch 'master' into yyjson

Commit:
add some more tests and remove FIXME because it's probably better this way

Commit:
make it possible to create default macro's elsewhere too

Commit:
more json functions for more sqlite/postgres compatibility

Commit:
some nitpicks, and json_group_structure macro

Commit:
more testing/fixing #3091

Commit:
Merge branch 'master' into yyjson

Commit:
try to make windows CI happy

Commit:
add some DUCKDB_API so windows is happy

Commit:
add some more DUCKDB_API

Commit:
add a bunch more DUCKDB_API for json extension

Commit:
remove unused import

lnkuiper commented:
This PR is ready for review, this time for real.

I've removed the typed extract functions in favour of a generic `json_extract` and `json_extract_string`, which are equivalent to postgres `json_extract_path` and `json_extract_path_text` (aliases added). These functions also have an operator, `->` and `->>` respectively.

Furthermore I've added various functions to create json from tabular data, which match functions in postgres/sqlite.

Finally, and most excitingly I've added functions that can "shred" json to create columnar data. The first function is `json_structure`:
```sql
select json_structure('[{"a":42}, {"a": null}]');
-- [{"a":"UBIGINT"}]
```

This function returns a "structure string", which can be used to "shred" JSON:
```sql
select json_transform('[{"a":42}, {"a": null}]', '[{"a":"UBIGINT"}]');
-- [{'a': 42}, {'a': NULL}]
```
The result is not JSON, but rather our own nested types, LIST and STRUCT. We can also cast to the desired type by specifying this in the structure string:
```sql
select json_transform('[{"a":42}, {"a": null}]', '[{"a":"HUGEINT"}]');
-- [{'a': 42}, {'a': NULL}]
```

We can now run TPC-H straight on JSON!
```sql
-- create json
create table lineitem_j as
    select to_json(row(l_orderkey,l_partkey,l_suppkey,l_linenumber,l_quantity,l_extendedprice,l_discount,l_tax,l_returnflag,
                       l_linestatus,l_shipdate,l_commitdate,l_receiptdate,l_shipinstruct,l_shipmode,l_comment)) as j
    from lineitem;
-- query 1 (note that we cast to INTEGER, DECIMAL, DATE, etc.
with transformed as (
    select json_transform(j, '{"l_orderkey":"INTEGER","l_partkey":"INTEGER","l_suppkey":"INTEGER","l_linenumber":"INTEGER","l_quantity":"INTEGER","l_extendedprice":"DECIMAL(15,2)","l_discount":"DECIMAL(15,2)","l_tax":"DECIMAL(15,2)","l_returnflag":"VARCHAR","l_linestatus":"VARCHAR","l_shipdate":"DATE","l_commitdate":"DATE","l_receiptdate":"DATE","l_shipinstruct":"VARCHAR","l_shipmode":"VARCHAR","l_comment":"VARCHAR"}') as j
    from lineitem_j
)
SELECT
    j.l_returnflag,
    j.l_linestatus,
    sum(j.l_quantity) AS sum_qty,
    sum(j.l_extendedprice) AS sum_base_price,
    sum(j.l_extendedprice * (1 - j.l_discount)) AS sum_disc_price,
    sum(j.l_extendedprice * (1 - j.l_discount) * (1 + j.l_tax)) AS sum_charge,
    avg(j.l_quantity) AS avg_qty,
    avg(j.l_extendedprice) AS avg_price,
    avg(j.l_discount) AS avg_disc,
    count(*) AS count_order
FROM
    transformed
WHERE
    j.l_shipdate <= CAST('1998-09-02' AS date)
GROUP BY
    j.l_returnflag,
    j.l_linestatus
ORDER BY
    j.l_returnflag,
    j.l_linestatus;
```

Q1 at SF1 on JSON takes around 2s on my laptop.

I forgot to mention that this JSON extension is built on [yyjson](https://github.com/ibireme/yyjson), which is a really nice JSON library written in C.

Commit:
Merge branch 'master' into yyjson

Commit:
remove microsoft-style identifiers from sqlite shell so our list notation works properly

Commit:
merge master

Mytherin commented on extension/json/json_common.cpp:
```diff
@@ -0,0 +1,135 @@
+	auto ptr = path_str.GetDataUnsafe();
+	// Empty strings and invalid $ paths yield an error
+	if (len == 0 || (*ptr == '$' && !JSONCommon::ValidPathDollar(ptr, len))) {
+		throw Exception("JSON path error");
```
Could we get a more descriptive error message perhaps, e.g. `JSONCommon::ValidPathDollar` could write to an error message parameter?

Commit:
add '%prec Op' to LAMBDA_ARROW

Mytherin commented on extension/json/json_functions/json_structure.cpp:
```diff
@@ -0,0 +1,186 @@
+	for (idx_t i = 1; i < elem_structures.size(); i++) {
+		auto elem_type = yyjson_mut_get_type(elem_structures[i]);
+		auto elem_type_string = yyjson_mut_get_str(elem_structures[i]);
+		if (type_string && strcmp(type_string, JSONCommon::TYPE_STRING_NULL) == 0) {
```
Do we need these string operations for type comparisons, or can we do this directly on the enums?

Mytherin commented on extension/json/json_functions/json_transform.cpp:
```diff
@@ -0,0 +1,375 @@
+static LogicalType StructureToTypeVal(yyjson_val *val) {
+	auto type_string = StringUtil::Lower(yyjson_get_str(val));
+	LogicalType result;
+	if ((StringUtil::StartsWith(type_string, "decimal(") || StringUtil::StartsWith(type_string, "dec(") ||
```
Do we need these string operations for type comparisons here, or is there a different way of reading this?

lnkuiper replied:
`TransformStringToLogicalType` returns  works for most logical types. The problem here is that `decimal` is parameterised. We do not have a standard way of parsing such strings into a logical type.

I could, for instance, change `TransformStringToLogicalType` to `TransformStringToLogicalTypeId`, and add a new function `TransformStringToLogicalType`, which can read parameterised types, and return an actual `LogicalType`, rather than `LogicalTypeId`. This is probably more future-proof.

Mytherin commented on src/storage/storage_info.cpp:
```diff
@@ -2,6 +2,6 @@
 namespace duckdb {
 
-const uint64_t VERSION_NUMBER = 31;
+const uint64_t VERSION_NUMBER = 32;
```
Is the storage version increment required?

Mytherin replied:
Only adding a new type should not break the storage version, could you try again and check what the error is (if there is still an error)?

Mytherin commented on test/extension/install_extension.test_slow:
```diff
@@ -110,6 +110,19 @@ SELECT id, first_name, last_name, email FROM PARQUET_SCAN('https://raw.githubuse
 10	Emily	Stewart	estewart9@opensource.org
 
+statement ok
+INSTALL 'json';
```
All of the code in this file is skipped after line 14, could you perhaps remove all of the code after the mode skip to prevent people from adding unreachable tests here in the future? We have a different mechanism of testing loading of extensions that runs the regular tests of each extension in the CI.

Mytherin reviewed (commented):
Thanks for the changes! Looks excellent. Some minor comments, otherwise this looks good to go from my side:

Commit:
implement some PR feedback

Commit:
attempt to fix memory leaks

Commit:
fix json_type

Commit:
decrement storage version again

Commit:
rename stringtologicaltype

Commit:
Merge branch 'master' into yyjson

Commit:
simplify json type casting

Commit:
progress on reworking json_structure

Commit:
code cleanup and more PR feedback

Commit:
add extensions to linux debug so we can see memory leaks etc.

Commit:
fix issue #3140

Commit:
Merge branch 'master' into yyjson

Commit:
make windows CI happy

Commit:
remove double env variable

Commit:
formatting

Commit:
funky unique_ptr for malloc'ed char*

Commit:
add default cases to some switches to make warnings go away

Commit:
cast to uint64_t so we don't get warnings

Commit:
remove ICU from CI (for now), and hopefully resolve windows CI issue

Commit:
merge master

Commit:
Merge branch 'master' into yyjson

Commit:
try to make win32 happy

Commit:
Merge branch 'master' into yyjson

Commit:
attempt to fix ICU memory leaks

Commit:
revert change to free functor because win32 segfaults anyway

Commit:
attempt to fix more icu leaks

Commit:
Merge branch 'master' into yyjson

Commit:
simplify json macro and formatting

Commit:
simplify json executors

Commit:
don't set vector type

Commit:
include missing header

Commit:
merge master

Commit:
attempt to fix win32 issues

Commit:
change yyjson iter loops, fix some bugs, try to make win32 happy

Commit:
Merge branch 'master' into yyjson

lnkuiper commented:
After a long wrestling match with a yyjson bug in Windows 32, I think this PR is now ready to be merged (let's wait for the CI first though!). I've circumvented the yyjson bug by using a different Iterator API for JSON objects and arrays, and it seemed to work on the windows machine that I was using for debugging, so it should work in the CI as well (fingers crossed!).

Quick recap for @Mytherin since he reviewed this last time. I've:
1. Implemented all the feedback
2. Cleaned up some code
3. Re-enabled extensions in the Linux Debug CI job
4. Fixed memory leaks in the JSON extension
5. ... fixed memory leaks in the ICU extension as well (had to do something while waiting for the slow windows machine to compile!)

I forgot to mention that this PR fixes issues #3070, #3091 and #3140.

lnkuiper commented:
Threadsan failure seems unrelated?

Mytherin commented:
Yes, codecov failure seems related though

Commit:
attempt to fix codecov ci failure

Mytherin commented on src/common/types.cpp:
```diff
@@ -560,13 +564,24 @@ LogicalTypeId TransformStringToLogicalType(const string &str) {
+	} else if (lower_str == "json") {
+		return LogicalTypeId::JSON;
+	} else if (lower_str == "null") {
+		return LogicalTypeId::SQLNULL;
```
This change now allows users to define the NULL type in e.g. create table statements, alter types, casts, etc. I'm not fundamentally against that, but this change would at least require testing NULL types in those locations that previously would not have allowed the null type to exist. Is this required for this PR, or is there another way of handling this functionality?

lnkuiper replied:
I've added some tests for this functionality in `test/sql/types/test_null_type.test`. Please let me know if this is sufficient or if more tests are needed.

Mytherin reviewed (approved):
Thanks for the updates! One last comment, then I think this is ready to merge:

Commit:
add test for NULL column type

Commit:
Merge branch 'master' into yyjson

Commit:
change map logicaltype tostring so it's the same as the sql definition

Mytherin commented:
Thanks! Looks great.
